README.md
<a href="https://covid19datahub.io"><img src="https://storage.covid19datahub.io/logo.svg" align="right" height="128"/></a>

# R Interface to COVID-19 Data Hub

![](https://www.r-pkg.org/badges/version/COVID19) ![](https://cranlogs.r-pkg.org/badges/last-month/COVID19) [![DOI](https://joss.theoj.org/papers/10.21105/joss.02376/status.svg)](https://doi.org/10.21105/joss.02376)

Provides a daily summary of COVID-19 cases, deaths, recovered, tests, vaccinations, and hospitalizations for 230+ countries, 760+ regions, and 12000+ administrative divisions of lower level.  Includes policy measures, mobility data, and geospatial identifiers. Data source: COVID-19 Data Hub https://covid19datahub.io

## Quickstart

```R
# install the package
install.packages("COVID19")

# load the package
library("COVID19")
```

## Usage

The only function in the package is `covid19()`. 

By default, the function downloads worldwide data by country:

```R
x <- covid19()
```

### Level

The argument `level` specifies the granularity of the data:

- 1: country-level data
- 2: state-level data
- 3: lower-level data

Download worldwide data by state:

```R
x <- covid19(level = 2)
```

### Country

The argument `country` filters the data by country. This is a vector of country names or [ISO codes](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes) (ISO 3166-1 Alpha-2 code, Alpha-3 code, or numeric code). 

Download data for Italy and Unites States at county/province level:

```R
x <- covid19(country = c("Italy", "US"), level = 3)
```

### Time range

The arguments `start` and `end` specify the period of interest. The data are subsetted to match this time range.

Download national-level data for United States from 01 October 2021 to 01 November 2021:

```R
x <- covid19("US", start = "2021-10-01", end = "2021-11-01")
```

### Vintage

The parameter `vintage` allows to retrieve the snapshot of the dataset that was available on the given date. This typically differs from subsetting the latest data, as most governments are updating the data retroactively. Available since 14 April, 2020.

Retrieve the data that were available on 15 May, 2020:

```R
x <- covid19(vintage = "2020-05-15")
```

### Download folder

The argument `dir` specifies the folder where the data files are to be downloaded. By default this is a temporary folder. 

Download the files in the folder `data`:

```R
dir.create("data")
x <- covid19(dir = "data")
```

### World Bank Open Data

Country-level covariates by [World Bank Open Data](https://data.worldbank.org/) can be added via the argument `wb`. This is a character vector of indicator codes to download. The codes can be found by inspecting the corresponding URL. For example, the code of the indicator "Hospital beds (per 1,000 people)" available at https://data.worldbank.org/indicator/SH.MED.BEDS.ZS is `SH.MED.BEDS.ZS`. The indicators are typically available at a yearly frequency. This function returns the latest data available between the `start` and the `end` date. See the table at the bottom of [this page](https://datatopics.worldbank.org/universal-health-coverage/coronavirus/) for suggested indicators. Example using GDP and number of hospital beds:

```R
x <- covid19(wb = c("gdp" = "NY.GDP.MKTP.CD", "hosp_beds" = "SH.MED.BEDS.ZS"))
```

### Google Mobility Reports

Mobility data by [Google Mobility Reports](https://www.google.com/covid19/mobility/) can be added via the argument `gmr`. This is the link to the Google "CSV by geographic area" ZIP folder. At the time of writing, the link is https://www.gstatic.com/covid19/mobility/Region_Mobility_Report_CSVs.zip. As the link has been stable since the beginning of the pandemic, the function accepts `gmr=TRUE` to automatically use this link.

```R
x <- covid19(gmr = TRUE)
```

### Apple Mobility Reports

Mobility data by [Apple Mobility Reports](https://covid19.apple.com/mobility) can be added via the argument `amr`. This is the link to the Apple "All CSV data" file. This link is changing constantly. Consider downloading the data file from the website first, and then set `amr="path/to/file.csv"`. If `amr=TRUE` is provided, the function tries to detect the latest URL from [this endpoint](https://covid19-static.cdn-apple.com/covid19-mobility-data/current/v3/index.json).

```R
x <- covid19(amr = TRUE)
```

## Documentation

See the full documentation [online](https://cran.r-project.org/package=COVID19/COVID19.pdf), or in R by typing `?covid19`

## Cite as

*Guidotti, E., Ardia, D., (2020), "COVID-19 Data Hub", Journal of Open Source Software 5(51):2376, doi: 10.21105/joss.02376.*

A BibTeX entry for LaTeX users is

```latex
@Article{,
    title = {COVID-19 Data Hub},
    year = {2020},
    doi = {10.21105/joss.02376},
    author = {Emanuele Guidotti and David Ardia},
    journal = {Journal of Open Source Software},
    volume = {5},
    number = {51},
    pages = {2376}
}
```


NEWS.md
# COVID19 3.0.2

* Argument `wb` now accepts also a non-named vector of indicators.

# COVID19 3.0.1

* Added option to automatically retrieve the URL to the Apple mobility CSV file. Thanks to [heinonmatti](https://github.com/covid19datahub/COVID19/issues/184)

# COVID19 3.0.0

* Compatible with [COVID-19 Data Hub v3](https://covid19datahub.io/news/index.html).



R/wb.R
worldbank <- function(x, indicator, start, end){
  
  if(!requireNamespace("wbstats", quietly = TRUE))
    stop("Package 'wbstats' needed but not installed. Install with: install.packages('wbstats')")
  
  # date
  start <- format(as.Date(start), "%Y")
  end   <- format(as.Date(end), "%Y")
  
  # download
  if(exists('wb_data', where = asNamespace('wbstats'), mode='function')){
    
    w <- wbstats::wb_data(indicator = indicator, start_date = start, end_date = end, return_wide = FALSE)
    w <- w %>% tidyr::pivot_wider(c('iso3c','date'), names_from = 'indicator_id', values_from = 'value')
    
  }
  else{
    
    w <- wbstats::wb(indicator = indicator, startdate = start, enddate = end)
    w <- w %>% tidyr::pivot_wider(c('iso3c','date'), names_from = 'indicatorID', values_from = 'value')
    
  }
  
  # column names
  map         <- c("iso_alpha_3" = "iso3c", "date" = "date", indicator)
  from        <- unname(map)
  to          <- names(map)
  miss        <- to == ""
  to[miss]    <- from[miss]
  colnames(w) <- sapply(colnames(w), function(x) to[which(from==x)])

  # get most recent value for each country
  w <- w %>% 
    
    dplyr::group_by_at('iso_alpha_3') %>% 
    
    dplyr::arrange_at('date') %>%
    
    dplyr::group_map(function(x, iso) {
      
      iso <- iso[[1]]
      
      idx <- ifelse(is.na(x), 1L, row(x))
      idx <- apply(idx, 2, max)
      
      x <- sapply(1:ncol(x), function(i) x[idx[i], i]) 
      x$iso_alpha_3 <- iso
      
      return(x)
    
    }) %>%
    
    dplyr::bind_rows()
  
  # drop year
  w$date <- NULL
  
  # merge
  x <- merge(x, w, by = 'iso_alpha_3', all.x = TRUE)
  
  # return
  return(x)
  
}



R/apple.R
apple <- function(x, level, url, dir, verbose){
  
  # check
  if(is.null(x$key_apple_mobility))
    return(x)
  
  # sanitize url
  if(is.logical(url)){
    if(!url) return(x)
    jurl <- "https://covid19-static.cdn-apple.com/covid19-mobility-data/current/v3/index.json"
    json <- readLines(jurl, warn = FALSE)
    base <- gsub('^(.*)basePath":"(.*?)"(.*)$', "\\2", json)
    csv <- gsub('^(.*)csvPath":"(.*?)"(.*)$', "\\2", json)
    url <- paste0("https://covid19-static.cdn-apple.com", base, csv)
  }
  
  # read
  a <- data.table::fread(url, encoding = "UTF-8", na.strings = "", header = TRUE, showProgress = verbose)
  
  # format
  id.vars <-  c("region", "sub-region", "transportation_type")
  measure.vars <- which(grepl("^\\d{4}-\\d{2}-\\d{2}$", colnames(a)))
  a <- suppressWarnings(data.table::melt(a, id.vars = id.vars, measure.vars = measure.vars, variable.name = "date"))
  a <- data.table::dcast(a, region + `sub-region` + date ~ transportation_type, value.var = "value")
  
  # date
  a$date <- as.Date(a$date)
  
  # key
  a$key_apple_mobility <- a$region
  idx <- which(!is.na(a$`sub-region`))
  a$key_apple_mobility[idx] <- paste(a$region[idx], a$`sub-region`[idx], sep = ", ")

  # subset
  a <- a[,c("date", "key_apple_mobility", "driving", "transit", "walking")]
  
  # return  
  join(x, a, on = c("date" = "date", "key_apple_mobility" = "key_apple_mobility"))

}


R/utils.R
# Site URL
baseurl <- "https://storage.covid19datahub.io"

# Build endpoint
endpoint <- function(...){
  ep <- paste0(list(...), collapse = "")
  paste(baseurl, ep, sep = "/")
}

# Generate local filename
local <- function(..., dir, timestamp){
  file <- paste0(list(...), collapse = "/")
  file <- gsub("^https?://[^/]*/", "", file)
  file <- gsub("\\.(gz|zip)$", "", file)
  if(timestamp) 
    dir <- paste(dir, format(Sys.time(), "%Y-%m-%d"), sep = "/")
  paste(dir, file, sep = "/")
}

# Read GZ files
read.gz <- function(url, dir, verbose){
  file <- download(url, dir = dir, verbose = verbose, timestamp = TRUE)
  data.table::fread(file, showProgress = verbose, encoding = "UTF-8", na.strings = "", keepLeadingZeros = TRUE)
}

# Read vintage ZIP
read.zip <- function(url, dir, level, verbose){
  file <- download(url, dir = dir, verbose = verbose, timestamp = FALSE)
  data.table::rbindlist(fill = TRUE, lapply(level, function(i){
    rawdata <- sprintf("%s/rawdata-%s.csv", file, i)
    dt <- try(data.table::fread(
      rawdata, showProgress = verbose, encoding = "UTF-8", na.strings = "", keepLeadingZeros = TRUE), 
      silent = !verbose)
    if("try-error" %in% class(dt)) return(NULL)
    return(dt)
  }))
}

# Read SQLite database
read.db <- function(url, dir, country, level, start, end, verbose){
  if(!requireNamespace("RSQLite", quietly = TRUE))
    stop("Package 'RSQLite' needed but not installed. Install with: install.packages('RSQLite')")
  level <- paste(level, collapse = "','")
  country <- paste(country, collapse = "','")
  sql <- sprintf("
    SELECT * 
    FROM 
      timeseries NATURAL JOIN location 
    WHERE 
      date BETWEEN '%s' AND '%s' AND
      administrative_area_level IN ('%s')", start, end, level)
  if(country!="")
    sql <- sprintf("%s AND (
        administrative_area_level_1 IN ('%s') OR
        iso_alpha_3 IN ('%s') OR
        iso_alpha_2 IN ('%s') OR
        iso_numeric IN ('%s')
      )", sql, country, country, country, country)
  file <- download(url, dir = dir, verbose = verbose, timestamp = FALSE)
  con <- RSQLite::dbConnect(RSQLite::SQLite(), file)
  x <- RSQLite::dbGetQuery(con, sql)
  RSQLite::dbDisconnect(con)
  return(x)
}

# Download and return local filename
download <- function(url, dir, verbose, timestamp){
  file <- local(url, dir = dir, timestamp = timestamp)
  if(file.exists(file)) 
    return(file)
  tmp <- tempfile()
  utils::download.file(url, destfile = tmp, mode = "wb", quiet = !verbose)
  ext <- tools::file_ext(url)
  if(ext=="gz")
    R.utils::gunzip(tmp, file)
  else if(ext=="zip")
    utils::unzip(tmp, exdir = file)
  else
    file <- NULL
  return(file)
}

# Subset the data by country, level, and time range
filter <- function(x, country, level, start, end){
  x <- x[x$date>=start & x$date<=end & x$administrative_area_level %in% level,]
  if(!is.null(country)){
    x <- x[x$administrative_area_level_1 %in% country |
           x$iso_alpha_3 %in% country |
           x$iso_alpha_2 %in% country |
           x$iso_numeric %in% country,]
  }
  return(x)
}

# Efficient left join
join <- function(x, y, on){
  data.table::setnames(y, old = on, new = names(on))
  data.table::setcolorder(y[x, on = names(on)], unique(names(x), names(y)))
}


R/google.R
google <- function(x, level, url, dir, verbose){
  
  # sanitize url
  if(is.logical(url)){
    if(!url) return(x)
    url <- "https://www.gstatic.com/covid19/mobility/Region_Mobility_Report_CSVs.zip"
  }
  
  # backward compatibility: use names instead of place_id
  backward <- any(nchar(x$key_google_mobility)!=27, na.rm = TRUE)
  if(backward){
  
    if(is.null(x$key_google_mobility)){
      idx <- which(x$administrative_area_level==1)
      if(!length(idx)) return(x)
      x$key_google_mobility[idx] <- x$iso_alpha_2[idx]
    }
    
    if(length(idx <- which(x$administrative_area_level>1))){
      x$key_google_mobility[idx] <- paste(x$iso_alpha_2[idx], x$key_google_mobility[idx], sep = ", ")
      x$key_google_mobility[idx] <- gsub("(, (NA)?)+$", "", x$key_google_mobility[idx])    
    }
    
  }

  # download
  path <- download(url, dir = dir, verbose = verbose, timestamp = TRUE)
  
  # files to read
  pattern <- sprintf("\\_%s\\_", paste0(unique(x$iso_alpha_2), collapse = "|"))
  files <- list.files(path, pattern = pattern, full.names = TRUE)
  
  # read
  g <- data.table::rbindlist(fill = TRUE, lapply(files, function(file){
    g <- data.table::fread(file, encoding = "UTF-8", na.strings = "", showProgress = verbose)
    if(backward){
      g$place_id <- paste(g$country_region_code, g$sub_region_1, g$sub_region_2, g$metro_area, sep = ", ")
      g$place_id <- gsub("(, (NA)?)+$", "", g$place_id)
    }
    g[g$place_id %in% x$key_google_mobility & !is.na(g$place_id),]
  }))
  
  # check
  if(!nrow(g))
    return(x)
  
  # convert date
  g$date <- as.Date(g$date)
  
  # subset
  g <- g[,c(
    "place_id", "date",            
    "retail_and_recreation_percent_change_from_baseline",
    "grocery_and_pharmacy_percent_change_from_baseline", 
    "parks_percent_change_from_baseline",    
    "transit_stations_percent_change_from_baseline",     
    "workplaces_percent_change_from_baseline",         
    "residential_percent_change_from_baseline")]
  
  # return
  join(x, g, on = c("date" = "date", "key_google_mobility" = "place_id"))
  
}


R/worldbank.R
worldbank <- function(x, indicator, start, end){
  
  # check
  if(!requireNamespace("wbstats", quietly = TRUE))
    stop("Package 'wbstats' needed but not installed. Install with: install.packages('wbstats')")
  
  # date
  start <- format(as.Date(start), "%Y")
  end   <- format(as.Date(end), "%Y")
  
  # set names
  if(is.null(names(indicator)))
    names(indicator) <- indicator
  
  # download
  w <- wbstats::wb_data(indicator = indicator, start_date = start, end_date = end, return_wide = FALSE)
  
  # convert to data.table and sort
  w <- data.table::data.table(w)
  data.table::setkey(w, iso3c, date)
  
  # bind variables for CRAN
  value <- iso3c <- indicator_id <- NULL
  # fill most recent value
  w[, value := data.table::nafill(value, type = "locf"), by = list(iso3c, indicator_id)]
  
  # pivot wider
  w <- data.table::dcast(w, iso3c + date ~ indicator_id, value.var = "value", fill = NA)
  
  # get most recent value for each country
  w <- unique(w, by = "iso3c", fromLast = TRUE)  
    
  # drop year
  w$date <- NULL
  
  # rename
  data.table::setnames(w, old = indicator, new = names(indicator))
  
  # return
  join(x, w, on = c("iso_alpha_3" = "iso3c"))
  
}


R/covid19cite.R
#' COVID-19 Data Hub Data Sources
#'
#' Retrieve citation and metadata of the data sources.
#' 
#' @param x dataset returned by \code{\link{covid19}}.
#' @param verbose logical. Print citation? Default \code{TRUE}. 
#'
#' @return \code{data.frame} of data sources.
#'
#' @examples
#' \dontrun{
#'
#' # Download data 
#' x <- covid19("USA")
#' 
#' # Data Sources
#' s <- covid19cite(x)
#' View(s)
#' }
#'
#' @export
#'
covid19cite <- function(x, verbose = TRUE){

  x <- attr(x, 'src')
  if(is.null(x))
    stop("No metadata found. Make sure 'x' is a dataset returned by the covid19() function.")
  
  if(verbose){
    
    y <- x %>%
      dplyr::distinct_at(c('title'), .keep_all = TRUE)
    
    y <- apply(y, 1, function(y){
      
      textVersion <- y['textVersion']
      if(is.na(textVersion))
        textVersion <- paste0(y['title'],' (',y['year'],')',', ',y['url'])
      
      utils::bibentry(
        bibtype     = ifelse(is.na(y['bibtype']), "Misc", y['bibtype']),
        title       = y['title'],
        year        = y['year'],
        author      = y['author'],
        institution = y['institution'],
        textVersion = textVersion
      )
      
    })
    
    for(i in 1:length(y)){
      if(i==1)
        cit <- y[[i]]
      else
        cit <- c(y[[i]], cit)
    }
      
    print(cit, style = "citation")
    
  }
  
  return(x)
  
}

R/covid19.R
#' COVID-19 Data Hub
#'
#' Download COVID-19 data from \url{https://covid19datahub.io}
#'
#' @param country vector of country names or \href{https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes}{ISO codes} (ISO 3166-1 Alpha-2 code, Alpha-3 code, or numeric code). By default, downloads data for all countries.
#' @param level integer. Granularity level. 1: country-level data. 2: state-level data. 3: lower-level data.
#' @param start,end the start and the end date of the period of interest. The data are subsetted to match this time range.
#' @param vintage date. This parameter allows to retrieve the snapshot of the dataset that was available on the given date. This typically differs from subsetting the latest data, as most governments are updating the data retroactively. Available since 2020-04-14.
#' @param wb character vector of \href{https://data.worldbank.org}{World Bank} indicator codes. See details.
#' @param gmr link to the \href{https://www.google.com/covid19/mobility/}{Google Mobility Report} dataset, or \code{TRUE}. See details.
#' @param amr link to the \href{https://covid19.apple.com/mobility}{Apple Mobility Report} dataset, or \code{TRUE}. See details.
#' @param dir folder where the data files are to be downloaded. 
#' @param verbose logical. Print on progress? Default \code{TRUE}. 
#' @param ... backward compatibility, not used.
#'
#' @details 
#' 
#' Country-level covariates by \href{https://data.worldbank.org}{World Bank Open Data} can be added via the argument \code{wb}.
#' This is a character vector of indicator codes to download.
#' The codes can be found by inspecting the corresponding URL. 
#' For example, the code of the indicator "Hospital beds (per 1,000 people)" available at \url{https://data.worldbank.org/indicator/SH.MED.BEDS.ZS} is \code{SH.MED.BEDS.ZS}.
#' The indicators are typically available at a yearly frequency. 
#' This function returns the latest data available between the \code{start} and the \code{end} date.
#' See the table at the bottom of \href{https://datatopics.worldbank.org/universal-health-coverage/coronavirus/}{this page} for suggested indicators.
#'
#' Mobility data by \href{https://www.google.com/covid19/mobility/}{Google Mobility Reports} can be added via the argument \code{gmr}.
#' This is the link to the Google "CSV by geographic area" ZIP folder. 
#' At the time of writing, the link is \url{https://www.gstatic.com/covid19/mobility/Region_Mobility_Report_CSVs.zip}.
#' As the link has been stable since the beginning of the pandemic, the function accepts \code{gmr=TRUE} to automatically use this link.
#' 
#' Mobility data by \href{https://covid19.apple.com/mobility}{Apple Mobility Reports} can be added via the argument \code{amr}.
#' This is the link to the Apple "All CSV data" file. This link is changing constantly. 
#' Consider downloading the data file from the website first, and then set \code{amr="path/to/file.csv"}.
#' If \code{amr=TRUE} is provided, the function tries to detect the latest URL from \href{https://covid19-static.cdn-apple.com/covid19-mobility-data/current/v3/index.json}{this endpoint}.
#'
#' Refer to \href{https://covid19datahub.io/reference/index.html}{this webpage} for the details on the data sources, and 
#' \href{https://covid19datahub.io/news/index.html}{see the changelog} for the latest news about the dataset.
#'
#' @return \code{data.frame}. See the \href{https://covid19datahub.io/articles/docs.html}{dataset documentation}
#'
#' @examples
#' \dontrun{
#'
#' # Worldwide data by country
#' x <- covid19()
#'
#' # Worldwide data by state
#' x <- covid19(level = 2)
#'
#' # Data for specific countries by county/province
#' x <- covid19(c("Italy", "US"), level = 3)
#' 
#' # Retrieve the data that were available on 15 May, 2020
#' x <- covid19(vintage = "2020-05-15")
#' 
#' # Download the files in the folder "data"
#' dir.create("data")
#' x <- covid19(dir = "data")
#' 
#' # World Bank data
#' wb <- c("gdp" = "NY.GDP.MKTP.CD", "hosp_beds" = "SH.MED.BEDS.ZS")
#' x  <- covid19(wb = wb)
#' 
#' # Google Mobility Reports
#' x <- covid19(gmr = TRUE)
#' 
#' # Apple Mobility Reports
#' x <- covid19(amr = TRUE)
#' 
#' }
#'
#' @source \url{https://covid19datahub.io}
#'
#' @references 
#' Guidotti, E., Ardia, D., (2020), "COVID-19 Data Hub", Journal of Open Source Software 5(51):2376, \doi{10.21105/joss.02376}.
#'
#' @note 
#' We have invested a lot of time and effort in creating \href{https://covid19datahub.io}{COVID-19 Data Hub}, please:
#' 
#' \itemize{
#' \item cite \href{https://joss.theoj.org/papers/10.21105/joss.02376}{Guidotti and Ardia (2020)} when using \href{https://covid19datahub.io}{COVID-19 Data Hub}.
#' \item place the URL \url{https://covid19datahub.io} in a footnote to help others find \href{https://covid19datahub.io}{COVID-19 Data Hub}.
#' \item you assume full risk for the use of \href{https://covid19datahub.io}{COVID-19 Data Hub}. 
#' We try our best to guarantee the data quality and consistency and the continuous filling of the Data Hub. 
#' However, it is free software and comes with ABSOLUTELY NO WARRANTY. 
#' Reliance on \href{https://covid19datahub.io}{COVID-19 Data Hub} for medical guidance or use of \href{https://covid19datahub.io}{COVID-19 Data Hub} in commerce is strictly prohibited.
#' }
#' 
#' @importFrom data.table :=
#' 
#' @export
#'
covid19 <- function(country = NULL,
                    level   = 1,
                    start   = "2010-01-01",
                    end     = Sys.Date(),
                    vintage = NULL,
                    wb      = NULL,
                    gmr     = NULL,
                    amr     = NULL,
                    dir     = tempdir(),
                    verbose = TRUE,
                    ...){

  if(any(!level %in% 1:3))
    stop("'level' must be one of 1, 2, 3 or a combination of those.")

  if(is.logical(vintage)){
    if(!vintage){
      vintage <- NULL
    }
    else{
      vintage <- end
      warning(sprintf("Using vintage='%s' (end date). See ?covid19 for the new usage of the 'vintage' parameter.", vintage))
    }
  }
  
  if(is.null(vintage)){
    
    if(is.null(country) | all(level==1)){
      x <- data.table::rbindlist(fill = TRUE, lapply(level, function(i){
        url <- endpoint("level/", i, ".csv.gz")
        read.gz(url, dir = dir, verbose = verbose)
      }))
      x <- filter(x, country = country, level = level, start = start, end = end)
    }
    
    else{
      url <- endpoint("country/index.csv.gz")
      map <- read.gz(url, dir = dir, verbose = verbose)
      iso <- map$iso_alpha_3[
        map$name %in% country |
        map$iso_alpha_3 %in% country |
        map$iso_alpha_2 %in% country |
        map$iso_numeric %in% country ]
      x <- data.table::rbindlist(fill = TRUE, lapply(iso, function(i){
        url <- endpoint("country/", i, ".csv.gz")
        read.gz(url, dir = dir, verbose = verbose)
      }))
      x <- filter(x, country = country, level = level, start = start, end = end)
    }
    
  }
  
  else{
    
    url <- sprintf("%s/%s%s", baseurl, vintage, ifelse(vintage>="2021-11-15", ".db.gz", ".zip"))
    ext <- tools::file_ext(url)
    if(ext=="zip"){
      x <- read.zip(url, dir = dir, level = level, verbose = verbose)  
      x <- filter(x, country = country, level = level, start = start, end = end)
    }
    if(ext=="gz"){
      x <- read.db(url, dir = dir, country = country, level = level, start, end, verbose = verbose)
    }
    
  }

  x$date <- as.Date(x$date)
  x <- data.table::data.table(x)
  
  if(!is.null(wb))
    x <- worldbank(x, indicator = wb, start = start, end = end)
  
  if(!is.null(gmr))
    x <- google(x, level = level, url = gmr, dir = dir, verbose = verbose)
  
  if(!is.null(amr))
    x <- apple(x, level = level, url = amr, dir = dir, verbose = verbose)
  
  if(verbose){
    cat("We have invested a lot of time and effort in creating COVID-19 Data Hub, please cite the following when using it:\n")
    print(utils::citation("COVID19"))
    cat("To hide this message use 'verbose = FALSE'.\n")
  }
  
  return(data.frame(x))
  
}
